producer/buffer/buffer.go

```diff
@@ -0,0 +1,197 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package buffer
+
+import (
+	"container/list"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3msg/producer"
+	"github.com/m3db/m3msg/producer/data"
+
+	"github.com/uber-go/atomic"
+	"github.com/uber-go/tally"
+)
+
+var (
+	errBufferFull                         = errors.New("buffer full")
+	errBufferClosed                       = errors.New("buffer closed")
+	errInvalidDataLargerThanMaxBufferSize = errors.New("invalid data, larger than max buffer size")
+)
+
+type bufferMetrics struct {
+	messageDropped  tally.Counter
+	bytesDropped    tally.Counter
+	messageBuffered tally.Gauge
+	bytesBuffered   tally.Gauge
+}
+
+func newBufferMetrics(scope tally.Scope) *bufferMetrics {
+	return &bufferMetrics{
+		messageDropped:  scope.Counter("message-dropped"),
+		bytesDropped:    scope.Counter("bytes-dropped"),
+		messageBuffered: scope.Gauge("message-buffered"),
+		bytesBuffered:   scope.Gauge("bytes-buffered"),
+	}
+}
+
+type buffer struct {
+	sync.Mutex
+
+	strategy        OnFullStrategy
+	maxBufferSize   uint64
+	cleanupInterval time.Duration
+	buffer          *list.List
+	m               *bufferMetrics
+
+	size     *atomic.Uint64
+	doneCh   chan struct{}
+	isClosed bool
+}
+
+// NewBuffer returns a new buffer.
+func NewBuffer(opts Options) producer.Buffer {
+	if opts == nil {
+		opts = NewBufferOptions()
+	}
+	return &buffer{
+		size:            atomic.NewUint64(0),
+		buffer:          list.New(),
+		isClosed:        false,
+		doneCh:          make(chan struct{}),
+		strategy:        opts.OnFullStrategy(),
+		maxBufferSize:   opts.MaxBufferSize(),
+		cleanupInterval: opts.CleanupInterval(),
+		m:               newBufferMetrics(opts.InstrumentOptions().MetricsScope()),
+	}
+}
+
+func (b *buffer) Buffer(d producer.Data) (producer.RefCountedData, error) {
+	b.Lock()
+	if b.isClosed {
+		b.Unlock()
+		return nil, errBufferClosed
+	}
+	var (
+		dataSize      = uint64(d.Size())
+		maxBufferSize = b.maxBufferSize
+	)
+	if dataSize > maxBufferSize {
+		b.Unlock()
+		return nil, errInvalidDataLargerThanMaxBufferSize
+	}
+	targetBufferSize := maxBufferSize - dataSize
```

Comment:
I may be missing something here but it seems like there might be a flaw here.  Lets say the `maxBufferSize` is 100 bytes and the `dataSize` is 80 bytes. It looks like here the `targetBufferSize` would be 20 bytes. Then we would call `dropEarliestUntilTargetWithLock` and free up 20 bytes, which would still not be enough room to fit the incoming data. 

Let me know if I am totally missing something here.

Reply:
`dropEarliestUntilTargetWithLock` will keep dropping data until the buffer hits the target size, basically it will keep drop earliest data until the buffer size reaches 20 in your example

Reply:
Sure but in my example the incoming data block is 80 bytes, so would we not want to free up 80 bytes rather than 20?

Reply:
The buffer will continue to drop data until it frees 80 bytes and hit 20.